### ISSUE TITLE
SC 25 | BurstEngine: An efficient distributed framework for training transformers On extremely Long sequences of over 1M tokens

### ISSUE THREAD

**Issue by woshipapa** [AI-Infra-Team/awesome-papers#174]
BurstEngine: An efficient distributed framework for training transformers On extremely Long sequences of over 1M tokens

[Ao Sun](https://dl.acm.org/doi/10.1145/3712285.3759802#), [Weilin Zhao](https://dl.acm.org/doi/10.1145/3712285.3759802#), [Xu Han](https://dl.acm.org/doi/10.1145/3712285.3759802#), [Cheng Yang](https://dl.acm.org/doi/10.1145/3712285.3759802#), [Zhiyuan Liu](https://dl.acm.org/doi/10.1145/3712285.3759802#), [Chuan Shi](https://dl.acm.org/doi/10.1145/3712285.3759802#), [Maosong Sun]

[Paper](https://dl.acm.org/doi/10.1145/3712285.3759802)

Motivation:
Although training Transformer-based models on long sequences is crucial for generating extensive outputs like documents and videos, existing distributed training methods (such as Tensor Parallelism and Context Parallelism) exhibit low Model FLOPs Utilization (MFU) as sequence lengths and GPU counts scale. BurstEngine is motivated by three core limitations. First, standard distributed attention mechanisms (like RingAttention) suffer from high communication overhead and poor communication-computation overlap, severely bottlenecking system throughput. Second, processing extremely long sequences incurs prohibitive memory consumption due to the need to store massive intermediate states, particularly in the attention modules and the Language Modeling (LM) head. Third, when applying necessary sparse attention patterns (such as causal or block-wise masking), distributed systems suffer from severe workload imbalance across GPUs, causing idle time that drags down overall efficiency. BurstEngine is therefore motivated by the need for a comprehensive, system-level distributed framework that jointly optimizes communication topology, memory management, and workload balance to efficiently train Transformers on sequences exceeding 1M tokens.

<img width="800" height="930" alt="Image" src="https://github.com/user-attachments/assets/818aeae5-626b-4e4d-800c-ff948e9e22fd" />

<img width="857" height="1280" alt="Image" src="https://github.com/user-attachments/assets/c7addf4d-04e9-4bd8-bdff-fbee7b05224d" />

BurstEngine introduces an optimized distributed attention mechanism that drastically reduces communication bottlenecks. It:

- reduces backward pass communication cost by nearly 25% by mathematically reformulating the gradient computation (passing Queries instead of Keys/Values).

<img width="895" height="534" alt="Image" src="https://github.com/user-attachments/assets/4da9fbfb-ef2c-47ac-a145-911937a8a390" />

- utilizes topology-aware ring communication to separate high-bandwidth intra-node (NVLink) and lower-bandwidth inter-node (InfiniBand) traffic.

- achieves fine-grained communication-computation overlap using a dedicated three-buffer swapping strategy for both activations and gradients.

Sequence-level Selective Checkpointing

To balance memory consumption and computational overhead in attention modules, BurstEngine introduces a novel gradient checkpointing scheme. It:
- divides the sequence into two segments based on recomputation cost.stores the second segment (which has higher recomputation overhead) directly in memory.recomputes only the first segment during the backward pass, cutting memory consumption by 50% with minimal throughput loss.

<img width="879" height="601" alt="Image" src="https://github.com/user-attachments/assets/9fa671b6-b952-4e94-8202-9b011ff7d364" />

Sequence-level Fusion of LM Head and Loss Function

To eliminate the memory disaster caused by storing massive Logits and Probability matrices at the end of the network, BurstEngine fuses the LM head with the cross-entropy loss function. It:
- splits the hidden states and vocabulary weights into smaller tiles.computes the local LogSumExp ($Lse$) and the loss without instantiating the full probability matrix.
- interleaves the forward and backward passes immediately at the tile level, allowing gradients to be computed on the fly without any recomputation overhead.

<img width="861" height="824" alt="Image" src="https://github.com/user-attachments/assets/7fe19527-cf57-4194-8633-924fee4feb6e" />

Sparse Attention Integration:

- To ensure all GPUs are fully utilized when using sparse attention masks, BurstEngine introduces fine-grained workload balancing strategies:
Zigzag-way balance: partitions the sequence so each GPU receives a front and a back subsequence, naturally complementing the heavy/light computational loads of causal masking.

- Striped-way balance: distributes tokens (or blocks of tokens) across GPUs in an interleaved, round-robin fashion, ensuring every GPU executes the exact same amount of compute for both causal and block-wise sparse attention.

<img width="1567" height="428" alt="Image" src="https://github.com/user-attachments/assets/fbb609c9-736f-4baf-9b53-d5c90f26de19" />
